osutil/sync_dir.go

```diff
@@ -0,0 +1,100 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// FileState describes the expected content and meta data of a single file.
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+}
+
+var errSameState = fmt.Errorf("file state has not changed")
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions are what is desired. Unexpected files are
+// removed. Missing files are created and differing files are corrected.  Files
+// not matching the pattern are ignored.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
+// returned to the caller for any extra processing that might be required (e.g.
+// to run some helper program).
+func EnsureDirState(dir, glob string, content map[string]*FileState) (changed, removed []string, err error) {
+	for baseName, fileState := range content {
+		filePath := filepath.Join(dir, baseName)
+		err := writeFile(filePath, fileState)
+		if err == errSameState {
+			continue
+		}
+		if err != nil {
+			return changed, removed, err
+		}
+		changed = append(changed, baseName)
+	}
+	matches, err := filepath.Glob(filepath.Join(dir, glob))
+	if err != nil {
+		return changed, removed, err
+	}
+	for _, filePath := range matches {
+		baseName := filepath.Base(filePath)
+		if content[baseName] != nil {
+			continue
+		}
+		err := os.Remove(filePath)
+		if err != nil {
+			return changed, removed, err
+		}
+		removed = append(removed, baseName)
+	}
+	return changed, removed, nil
+}
+
+func writeFile(filePath string, fileState *FileState) error {
+	stat, err := os.Stat(filePath)
+	if err != nil {
```

Comment:
This error check says that if there is an error, go ahead and write the file, but this only makes sense if the error is for the basename of the filePath does not exist. What if another path component doesn't exist? What if the error is EPERM (which can happen with more than just DAC perms). 'man fstatat' gives a lot of possible failure conditions. Looking at https://golang.org/src/os/file_unix.go?s=4060:4099#L150 it appears go itself is not being very helpful here, simply returning PathError if anything goes wrong.